pkg/aws_keyhub/saml.go

```diff
@@ -0,0 +1,112 @@
+package aws_keyhub
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"encoding/xml"
+	"github.com/sirupsen/logrus"
+	"strings"
+)
+
+func DecodeSAMLResponse(samlResponse string) []byte {
+	decoded, err := base64.URLEncoding.DecodeString(samlResponse)
+	if err != nil {
+		logrus.Println("Failed to decode SAML Response from KeyHub.")
+		logrus.Fatal(err)
+	}
+
+	logrus.Debugln("Decoded SAML Response", string(decoded))
+	return decoded
+}
+
+func RolesAndPrincipalsFromSamlResponse(samlResponseDecoded []byte) map[string]RolesAndPrincipals {
+	var response Response
+	xml.Unmarshal(samlResponseDecoded, &response)
+	logrus.Debugln("Unmarshalled SAML Assertion:", response)
+
+	var rolesAndPrincipals = make(map[string]RolesAndPrincipals)
+
+	for _, assertion := range response.Assertion {
+		for _, attributeStatement := range assertion.AttributeStatement {
+			for _, attribute := range attributeStatement.Attribute {
+				for _, attributeValue := range attribute.AttributeValue {
+
+					if attribute.Name == "https://github.com/topicuskeyhub/aws-keyhub/groups" {
+
+						var groupsMetadata GroupsMetadata
+						if err := json.Unmarshal([]byte(attributeValue), &groupsMetadata); err != nil {
+							logrus.Warning("Cannot unmarshal JSON nested in the SAML Assertion containing groups metadata.")
+						}
+
+						if existingItem, exists := rolesAndPrincipals[groupsMetadata.Arn]; exists {
+							existingItem.Description = groupsMetadata.Description
+							rolesAndPrincipals[groupsMetadata.Arn] = existingItem
+						} else {
+							rolesAndPrincipals[groupsMetadata.Arn] = RolesAndPrincipals{Description: groupsMetadata.Description}
+						}
+					}
+
+					if attribute.Name == "https://aws.amazon.com/SAML/Attributes/Role" {
+						role, principal := splitRoleAndPrincipal(attributeValue)
+
+						if existingItem, exists := rolesAndPrincipals[attributeValue]; exists {
+							existingItem.Role = role
```

Comment:
zit deze role and principle ook al niet in de keyhub groups info?

Reply:
Dat klopt. Maar ik wil graag de functionaliteit van aws-keyhub werkend houden zonder het custom SAML attribuut. Oftewel als je een AWS compliant SAML koppeling hebt kun je inloggen d.m.v. aws-keyhub. Dan heb je wel iets minder informatie, maar kun je probleemlos inloggen, zie screenshot.
<img width="977" alt="Screenshot 2021-09-21 at 10 55 20" src="https://user-images.githubusercontent.com/981177/134144008-98b5bfc0-9ba1-4a6a-a288-ae0a611068bf.png">